Limit User Folders

A user managed to synchronise our server with their UWash IMAP
server's view of a very busy home directory, creating tens of
thousands of folders and showing that there's a DOS risk against
mailboxes.db here.

This patch creates a new config variable user_folder_limit which,
if set, returns IMAP_PERMISSION_DENIED to any further attempts to
create a folder if the user has too many folders already.

REQUIRES: cyrus-findall-txn-2.3.12.diff, cyrus-fastrename-2.3.12.diff

UPDATED: now uses the new count_inferiors which takes a transaction
ID, so is transaction safe even on old cyruses with dangerous
skiplist implementations.
Index: cyrus-imapd-2.3.12rc2/imap/mboxlist.c
===================================================================
--- cyrus-imapd-2.3.12rc2.orig/imap/mboxlist.c	2008-04-19 21:37:20.000000000 +1000
+++ cyrus-imapd-2.3.12rc2/imap/mboxlist.c	2008-04-19 21:37:57.000000000 +1000
@@ -370,6 +370,7 @@
     unsigned long parentpartitionlen = 0;
     unsigned long parentacllen = 0;
     int mbtype;
+    int user_folder_limit;
     
     /* Check for invalid name/partition */
     if (partition && strlen(partition) > MAX_PARTITION_LEN) {
@@ -443,6 +444,20 @@
 	    break;
 	}
     }
+
+    /* check the folder limit */
+    if (!isadmin && (user_folder_limit = config_getint(IMAPOPT_USER_FOLDER_LIMIT))) {
+	char buf[MAX_MAILBOX_NAME+1];
+	strncpy(buf, mbox, strlen(mbox));
+	if (!strncmp(buf, "user.", 5)) {
+	    char *firstdot = strchr(buf+5, '.');
+	    if (firstdot) *firstdot = '\0';
+	    if (mboxlist_count_inferiors(buf, isadmin, userid, auth_state, tid) + 1
+					 >= user_folder_limit)
+		return IMAP_PERMISSION_DENIED;
+	}
+    }
+
     if (parentlen != 0) {
 	/* check acl */
 	if (!isadmin &&
Index: cyrus-imapd-2.3.12rc2/lib/imapoptions
===================================================================
--- cyrus-imapd-2.3.12rc2.orig/lib/imapoptions	2008-04-19 21:36:06.000000000 +1000
+++ cyrus-imapd-2.3.12rc2/lib/imapoptions	2008-04-19 21:37:24.000000000 +1000
@@ -1056,6 +1056,10 @@
 { "umask", "077", STRING }
 /* The umask value used by various Cyrus IMAP programs. */
 
+{ "user_folder_limit", 0, INT }
+/* Limit the number of folders a user can create in their INBOX.  
+   Set to 0 (default) for no limit.  Only affects folders in user. */
+
 { "username_tolower", 1, SWITCH }
 /* Convert usernames to all lowercase before login/authenticate.  This
    is useful with authentication backends which ignore case during
